### ISSUE TITLE
Validation Accuracy fluctuating

### ISSUE THREAD

**Issue by heiznberg** [heiznberg/Deep-Learning-3D-CNN-Question#1]
Hey guys, I'm new to Github and also new to Deep Learning. I'm trying to implement a classifcation of brain MRI Scans in Nifti format to predict whether a patient has Parkinsons Disease or not. I have about 600 preprocessed MRI Scans (300 PD/300 Healthy Control) which I split in 70/30 train/validation data. My problem is that the val_acc and val_loss have good epochs where they reach at max ~81% validation accuracy and a good val_loss but then drop back to 50% validation accuracy for a few epochs. It is fluctuating very strongly. Meanwhile the training loss and accuracy are steadily decreasing/increasing as they should. The model with 81% accuracy was saved in a .h5 file and I made predictions with the saved model on files from the train/validation set. The accuracy seemed to be correct since the label was predicted correctly on most of these pictures. My code is based on this tutorial from the keras website: https://keras.io/examples/vision/3D_image_classification/

I have tried almost anything I could think of: changed network structure, changed hyperparameters (learning rate, dropout). Would it make sense to try different optimizers or maybe L1/L2 regularization? If yes, which one of these would make most sense to implement and in which point of the network? Below is my code. Thanks in advance

```
import os
import zipfile
import numpy as np
import tensorflow as tf
import random
import matplotlib.pyplot as plt
from tensorflow import keras
from tensorflow.keras import layers
import datetime
import nibabel as nib


from scipy import ndimage


def read_nifti_file(filepath):
    # read file
    scan = nib.load(filepath)
    # get raw data
    scan = scan.get_fdata()
    return scan

def normalize(volume):
    """Normalize the Volume"""
    min = 1
    max = 150
    volume[volume < min] = min
    volume[volume > max] = max
    volume = (volume - min) / (max - min)
    return volume

def resize_volume(img):
    """Resize across z-axis"""
    # set desired depth
    desired_depth = 64
    desired_width = 128
    desired_height = 128
    # get current depth
    current_depth = img.shape[-1]
    current_width = img.shape[0]
    current_height = img.shape[1]
    # compute depth factor
    depth = current_depth / desired_depth
    width = current_width / desired_width
    height = current_height / desired_height
    depth_factor = 1 / depth
    width_factor = 1 / width
    height_factor = 1 / height
    # Rotate
    # resize across z-axis
    img = ndimage.zoom(img, (width_factor, height_factor, depth_factor), order=1)
    return img

def remove_dimension(array):
    for elem in array:
        if elem.ndim == 4:
            np.squeeze(elem)
            return elem
        else:
            return elem
    return array


def process_scan_3d(path):
    """Read and resize Volume"""
    # Read scan
    volume = read_nifti_file(path)
    # Normalize
    # resize w,h and d
    volume = normalize(volume)
    volume = resize_volume(volume)
    return volume
# scans vom 1. download

DATADIR = "/media/Chaitu_storage/vladiWorkingSpace/"
CATEGORY1 = ["prep_healthy"]
CATEGORY2 = ["prep_pds"]

healthy_data = []
pd_data = []

def create_healthy_data():
    for category in CATEGORY1:
        path = os.path.join(DATADIR, category)  # pfad zu MRT Scans
        class_num = 0
        for img in os.listdir(path):
            nifti_array = process_scan_3d(os.path.join(path, img))
            healthy_data.append([nifti_array, class_num])

def create_pd_data():
    for category in CATEGORY2:
        path = os.path.join(DATADIR, category)  # pfad zu MRT Scans
        class_num = 1
        for img in os.listdir(path):
            nifti_array = process_scan_3d(os.path.join(path, img))
            pd_data.append([nifti_array, class_num])

create_healthy_data()
create_pd_data()


normal_scans = [] #pictures
normal_labels = [] #labels
abnormal_labels = []
abnormal_scans = []

for features, label in healthy_data:
    normal_scans.append(features)
    normal_labels.append(label)

for features, label in pd_data:
    abnormal_scans.append(features)
    abnormal_labels.append(label)

abnormal_scan_paths = [
    os.path.join(os.getcwd(), "/media/Chaitu_storage/vladiWorkingSpace/prep_pds/", x)
    for x in os.listdir("/media/Chaitu_storage/vladiWorkingSpace/prep_pds/")
]

normal_scan_paths = [
    os.path.join(os.getcwd(), "/media/Chaitu_storage/vladiWorkingSpace/prep_healthy/", x)
    for x in os.listdir("/media/Chaitu_storage/vladiWorkingSpace/prep_healthy/")
]

print("Healthy MRI Scans" + str(len(normal_scan_paths)))
print("PD MRI Scans" + str(len(abnormal_scan_paths)))

normal_scans = np.array(normal_scans)
normal_labels = np.array(normal_labels)
abnormal_scans = np.array(abnormal_scans)
abnormal_labels = np.array(abnormal_labels)


# Split data in the ratio 50-50 for training and validation.
x_train = np.concatenate((abnormal_scans[70:], normal_scans[70:]), axis=0)
y_train = np.concatenate((abnormal_labels[70:], normal_labels[70:]), axis=0)
x_val = np.concatenate((abnormal_scans[:70], normal_scans[:70]), axis=0)
y_val = np.concatenate((abnormal_labels[:70], normal_labels[:70]), axis=0)

print(
    "Number of samples in train and validation are %d and %d."
    % (x_train.shape[0], x_val.shape[0]),
)

def rotate(volume):

    def scipy_rotate(volume):
        #define rotation angles
        angles = [-20, -10, 5, 5, 10, 20]
        #pick angles at random
        angle = random.choice(angles)
        #rotate vol.
        volume = ndimage.rotate(volume, angle, reshape=False)
        volume[volume < 0] = 0
        volume[volume > 1] = 1
        return volume

    augmented_volume = tf.numpy_function(scipy_rotate, [volume], tf.float32)
    return augmented_volume

def train_preprocessing(volume, label):
    """Process training data by rotating and adding a channel"""
    #Rotate volume

    volume = tf.expand_dims(volume, axis=3)
    return volume, label

def validation_preprocessing(volume, label):
    """Process validation data by only adding a channel."""
    volume = tf.expand_dims(volume, axis=3)
    return volume, label

train_loader = tf.data.Dataset.from_tensor_slices((x_train, y_train)) #-> input muss Tensor sein
validation_loader = tf.data.Dataset.from_tensor_slices((x_val, y_val))

batch_size = 5


#augment on the fly during training
train_dataset = (
    train_loader.shuffle(len(x_train))
    .map(train_preprocessing)
    .batch(batch_size)
    .prefetch(2)
)

validation_dataset = (
    validation_loader.shuffle(len(x_val))
    .map(validation_preprocessing)
    .batch(batch_size)
    .prefetch(2)
)


def get_model(width=128, height=128, depth=64):
    """Build a 3D convolutional neural network model."""

    inputs = keras.Input((width, height, depth, 1))

    x = layers.Conv3D(filters=64, kernel_size=3, activation="relu")(inputs)
    x = layers.MaxPool3D(pool_size=2)(x)
    x = layers.BatchNormalization()(x)

    x = layers.Conv3D(filters=64, kernel_size=3, activation="relu")(x)
    x = layers.MaxPool3D(pool_size=2)(x)
    x = layers.BatchNormalization()(x)

    x = layers.Conv3D(filters=128, kernel_size=3, activation="relu")(x)
    x = layers.MaxPool3D(pool_size=2)(x)
    x = layers.BatchNormalization()(x)

    x = layers.Conv3D(filters=256, kernel_size=3, activation="relu")(x)
    x = layers.MaxPool3D(pool_size=2)(x)
    x = layers.BatchNormalization()(x)

    x = layers.GlobalAveragePooling3D()(x)
    x = layers.Dense(units=512, activation="relu")(x)
    x = layers.Dropout(0.3)(x)

    outputs = layers.Dense(units=1, activation="sigmoid")(x)

    # Define the model.
    model = keras.Model(inputs, outputs, name="3dcnn")
    return model

# Build model
model = get_model(width=128, height=128, depth=64)
model.load_weights("parkinson_pred_128.h5")
model.summary()

)
model.compile(
    loss="binary_crossentropy",
    optimizer=keras.optimizers.Adam(learning_rate=0.0001), #Change next run
    metrics=["acc"],
)


log_dir = "logs/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)




# Define callbacks.
checkpoint_cb = keras.callbacks.ModelCheckpoint(
    save_weights_only=False, filepath=".h5", save_best_only=True, monitor="val_acc"
)

# Train the model, doing validation at the end of each epoch
epochs = 100
model.fit(
    train_dataset,
    validation_data=validation_dataset,
    epochs=epochs,
    shuffle=True,
    verbose=2,
    callbacks=[checkpoint_cb,tensorboard_callback]
)

```

**Comment by Savant-HO** [comment]
> Hey guys, I'm new to Github and also new to Deep Learning. I'm trying to implement a classifcation of brain MRI Scans in Nifti format to predict whether a patient has Parkinsons Disease or not. I have about 600 preprocessed MRI Scans (300 PD/300 Healthy Control) which I split in 70/30 train/validation data. My problem is that the val_acc and val_loss have good epochs where they reach at max ~81% validation accuracy and a good val_loss but then drop back to 50% validation accuracy for a few epochs. It is fluctuating very strongly. Meanwhile the training loss and accuracy are steadily decreasing/increasing as they should. The model with 81% accuracy was saved in a .h5 file and I made predictions with the saved model on files from the train/validation set. The accuracy seemed to be correct since the label was predicted correctly on most of these pictures. My code is based on this tutorial from the keras website: https://keras.io/examples/vision/3D_image_classification/
> 
> I have tried almost anything I could think of: changed network structure, changed hyperparameters (learning rate, dropout). Would it make sense to try different optimizers or maybe L1/L2 regularization? If yes, which one of these would make most sense to implement and in which point of the network? Below is my code. Thanks in advance
> 
> ```
> import os
> import zipfile
> import numpy as np
> import tensorflow as tf
> import random
> import matplotlib.pyplot as plt
> from tensorflow import keras
> from tensorflow.keras import layers
> import datetime
> import nibabel as nib
> 
> 
> from scipy import ndimage
> 
> 
> def read_nifti_file(filepath):
>     # read file
>     scan = nib.load(filepath)
>     # get raw data
>     scan = scan.get_fdata()
>     return scan
> 
> def normalize(volume):
>     """Normalize the Volume"""
>     min = 1
>     max = 150
>     volume[volume < min] = min
>     volume[volume > max] = max
>     volume = (volume - min) / (max - min)
>     return volume
> 
> def resize_volume(img):
>     """Resize across z-axis"""
>     # set desired depth
>     desired_depth = 64
>     desired_width = 128
>     desired_height = 128
>     # get current depth
>     current_depth = img.shape[-1]
>     current_width = img.shape[0]
>     current_height = img.shape[1]
>     # compute depth factor
>     depth = current_depth / desired_depth
>     width = current_width / desired_width
>     height = current_height / desired_height
>     depth_factor = 1 / depth
>     width_factor = 1 / width
>     height_factor = 1 / height
>     # Rotate
>     # resize across z-axis
>     img = ndimage.zoom(img, (width_factor, height_factor, depth_factor), order=1)
>     return img
> 
> def remove_dimension(array):
>     for elem in array:
>         if elem.ndim == 4:
>             np.squeeze(elem)
>             return elem
>         else:
>             return elem
>     return array
> 
> 
> def process_scan_3d(path):
>     """Read and resize Volume"""
>     # Read scan
>     volume = read_nifti_file(path)
>     # Normalize
>     # resize w,h and d
>     volume = normalize(volume)
>     volume = resize_volume(volume)
>     return volume
> # scans vom 1. download
> 
> DATADIR = "/media/Chaitu_storage/vladiWorkingSpace/"
> CATEGORY1 = ["prep_healthy"]
> CATEGORY2 = ["prep_pds"]
> 
> healthy_data = []
> pd_data = []
> 
> def create_healthy_data():
>     for category in CATEGORY1:
>         path = os.path.join(DATADIR, category)  # pfad zu MRT Scans
>         class_num = 0
>         for img in os.listdir(path):
>             nifti_array = process_scan_3d(os.path.join(path, img))
>             healthy_data.append([nifti_array, class_num])
> 
> def create_pd_data():
>     for category in CATEGORY2:
>         path = os.path.join(DATADIR, category)  # pfad zu MRT Scans
>         class_num = 1
>         for img in os.listdir(path):
>             nifti_array = process_scan_3d(os.path.join(path, img))
>             pd_data.append([nifti_array, class_num])
> 
> create_healthy_data()
> create_pd_data()
> 
> 
> normal_scans = [] #pictures
> normal_labels = [] #labels
> abnormal_labels = []
> abnormal_scans = []
> 
> for features, label in healthy_data:
>     normal_scans.append(features)
>     normal_labels.append(label)
> 
> for features, label in pd_data:
>     abnormal_scans.append(features)
>     abnormal_labels.append(label)
> 
> abnormal_scan_paths = [
>     os.path.join(os.getcwd(), "/media/Chaitu_storage/vladiWorkingSpace/prep_pds/", x)
>     for x in os.listdir("/media/Chaitu_storage/vladiWorkingSpace/prep_pds/")
> ]
> 
> normal_scan_paths = [
>     os.path.join(os.getcwd(), "/media/Chaitu_storage/vladiWorkingSpace/prep_healthy/", x)
>     for x in os.listdir("/media/Chaitu_storage/vladiWorkingSpace/prep_healthy/")
> ]
> 
> print("Healthy MRI Scans" + str(len(normal_scan_paths)))
> print("PD MRI Scans" + str(len(abnormal_scan_paths)))
> 
> normal_scans = np.array(normal_scans)
> normal_labels = np.array(normal_labels)
> abnormal_scans = np.array(abnormal_scans)
> abnormal_labels = np.array(abnormal_labels)
> 
> 
> # Split data in the ratio 50-50 for training and validation.
> x_train = np.concatenate((abnormal_scans[70:], normal_scans[70:]), axis=0)
> y_train = np.concatenate((abnormal_labels[70:], normal_labels[70:]), axis=0)
> x_val = np.concatenate((abnormal_scans[:70], normal_scans[:70]), axis=0)
> y_val = np.concatenate((abnormal_labels[:70], normal_labels[:70]), axis=0)
> 
> print(
>     "Number of samples in train and validation are %d and %d."
>     % (x_train.shape[0], x_val.shape[0]),
> )
> 
> def rotate(volume):
> 
>     def scipy_rotate(volume):
>         #define rotation angles
>         angles = [-20, -10, 5, 5, 10, 20]
>         #pick angles at random
>         angle = random.choice(angles)
>         #rotate vol.
>         volume = ndimage.rotate(volume, angle, reshape=False)
>         volume[volume < 0] = 0
>         volume[volume > 1] = 1
>         return volume
> 
>     augmented_volume = tf.numpy_function(scipy_rotate, [volume], tf.float32)
>     return augmented_volume
> 
> def train_preprocessing(volume, label):
>     """Process training data by rotating and adding a channel"""
>     #Rotate volume
> 
>     volume = tf.expand_dims(volume, axis=3)
>     return volume, label
> 
> def validation_preprocessing(volume, label):
>     """Process validation data by only adding a channel."""
>     volume = tf.expand_dims(volume, axis=3)
>     return volume, label
> 
> train_loader = tf.data.Dataset.from_tensor_slices((x_train, y_train)) #-> input muss Tensor sein
> validation_loader = tf.data.Dataset.from_tensor_slices((x_val, y_val))
> 
> batch_size = 5
> 
> 
> #augment on the fly during training
> train_dataset = (
>     train_loader.shuffle(len(x_train))
>     .map(train_preprocessing)
>     .batch(batch_size)
>     .prefetch(2)
> )
> 
> validation_dataset = (
>     validation_loader.shuffle(len(x_val))
>     .map(validation_preprocessing)
>     .batch(batch_size)
>     .prefetch(2)
> )
> 
> 
> def get_model(width=128, height=128, depth=64):
>     """Build a 3D convolutional neural network model."""
> 
>     inputs = keras.Input((width, height, depth, 1))
> 
>     x = layers.Conv3D(filters=64, kernel_size=3, activation="relu")(inputs)
>     x = layers.MaxPool3D(pool_size=2)(x)
>     x = layers.BatchNormalization()(x)
> 
>     x = layers.Conv3D(filters=64, kernel_size=3, activation="relu")(x)
>     x = layers.MaxPool3D(pool_size=2)(x)
>     x = layers.BatchNormalization()(x)
> 
>     x = layers.Conv3D(filters=128, kernel_size=3, activation="relu")(x)
>     x = layers.MaxPool3D(pool_size=2)(x)
>     x = layers.BatchNormalization()(x)
> 
>     x = layers.Conv3D(filters=256, kernel_size=3, activation="relu")(x)
>     x = layers.MaxPool3D(pool_size=2)(x)
>     x = layers.BatchNormalization()(x)
> 
>     x = layers.GlobalAveragePooling3D()(x)
>     x = layers.Dense(units=512, activation="relu")(x)
>     x = layers.Dropout(0.3)(x)
> 
>     outputs = layers.Dense(units=1, activation="sigmoid")(x)
> 
>     # Define the model.
>     model = keras.Model(inputs, outputs, name="3dcnn")
>     return model
> 
> # Build model
> model = get_model(width=128, height=128, depth=64)
> model.load_weights("parkinson_pred_128.h5")
> model.summary()
> 
> )
> model.compile(
>     loss="binary_crossentropy",
>     optimizer=keras.optimizers.Adam(learning_rate=0.0001), #Change next run
>     metrics=["acc"],
> )
> 
> 
> log_dir = "logs/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
> tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
> 
> 
> 
> 
> # Define callbacks.
> checkpoint_cb = keras.callbacks.ModelCheckpoint(
>     save_weights_only=False, filepath=".h5", save_best_only=True, monitor="val_acc"
> )
> 
> # Train the model, doing validation at the end of each epoch
> epochs = 100
> model.fit(
>     train_dataset,
>     validation_data=validation_dataset,
>     epochs=epochs,
>     shuffle=True,
>     verbose=2,
>     callbacks=[checkpoint_cb,tensorboard_callback]
> )
> ```

Hello, I have met this problem too. Could you please tell me how to solve it? Thanks a lot~
